File: tools/schedule/schedule.go
package schedule

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"

	oasis "github.com/nevindra/oasis"
)

// Tool manages scheduled/recurring actions.
type Tool struct {
	store    oasis.Store
	tzOffset int // hours from UTC (e.g. 7 for WIB)
}

// New creates a ScheduleTool.
func New(store oasis.Store, tzOffset int) *Tool {
	return &Tool{store: store, tzOffset: tzOffset}
}

func (t *Tool) Definitions() []oasis.ToolDefinition {
	return []oasis.ToolDefinition{
		{
			Name:        "schedule_create",
			Description: "Create a scheduled/recurring action that runs automatically. Use when the user wants something done periodically (daily briefings, recurring searches, regular summaries).",
			Parameters: json.RawMessage(`{"type":"object","properties":{
				"description":{"type":"string","description":"Human-readable description of what this scheduled action does"},
				"time":{"type":"string","description":"Time in HH:MM format (24-hour, user's local timezone)"},
				"recurrence":{"type":"string","enum":["once","daily","custom","weekly","monthly"],"description":"How often to run"},
				"day":{"type":"string","description":"For weekly: day name. For custom: comma-separated day names. For monthly: day number (1-31)."},
				"tools":{"type":"array","items":{"type":"object","properties":{"tool":{"type":"string"},"params":{"type":"object"}},"required":["tool","params"]},"description":"Tools to execute when the schedule fires"},
				"synthesis_prompt":{"type":"string","description":"How to format/summarize results"}
			},"required":["description","time","recurrence","tools"]}`),
		},
		{
			Name:        "schedule_list",
			Description: "List all scheduled actions with their schedules, status, and next run time.",
			Parameters:  json.RawMessage(`{"type":"object","properties":{}}`),
		},
		{
			Name:        "schedule_update",
			Description: "Update a scheduled action: enable/disable it or change its schedule.",
			Parameters: json.RawMessage(`{"type":"object","properties":{
				"description_query":{"type":"string","description":"Substring to match the scheduled action description"},
				"enabled":{"type":"boolean","description":"Set to true to enable, false to disable/pause"},
				"time":{"type":"string","description":"New time in HH:MM format (optional)"},
				"recurrence":{"type":"string","enum":["once","daily","custom","weekly","monthly"],"description":"New recurrence (optional)"},
				"day":{"type":"string","description":"New day(s) (optional)"}
			},"required":["description_query"]}`),
		},
		{
			Name:        "schedule_delete",
			Description: "Delete a scheduled action. Matches by description substring, or '*' to delete all.",
			Parameters: json.RawMessage(`{"type":"object","properties":{
				"description_query":{"type":"string","description":"Substring to match the description, or '*' for all"}
			},"required":["description_query"]}`),
		},
	}
}

func (t *Tool) Execute(ctx context.Context, name string, args json.RawMessage) (oasis.ToolResult, error) {
	var result string
	var err error

	switch name {
	case "schedule_create":
		result, err = t.handleCreate(ctx, args)
	case "schedule_list":
		result, err = t.handleList(ctx)
	case "schedule_update":
		result, err = t.handleUpdate(ctx, args)
	case "schedule_delete":
		result, err = t.handleDelete(ctx, args)
	default:
		return oasis.ToolResult{Error: "unknown schedule tool: " + name}, nil
	}

	if err != nil {
		return oasis.ToolResult{Error: err.Error()}, nil
	}
	return oasis.ToolResult{Content: result}, nil
}

func (t *Tool) handleCreate(ctx context.Context, args json.RawMessage) (string, error) {
	var p struct {
		Description     string          `json:"description"`
		Time            string          `json:"time"`
		Recurrence      string          `json:"recurrence"`
		Day             string          `json:"day"`
		Tools           json.RawMessage `json:"tools"`
		SynthesisPrompt string          `json:"synthesis_prompt"`
	}
	if err := json.Unmarshal(args, &p); err != nil {
		return "", fmt.Errorf("invalid args: %w", err)
	}

	schedule := buildScheduleString(p.Time, p.Recurrence, p.Day)
	now := oasis.NowUnix()
	nextRun, ok := ComputeNextRun(schedule, now, t.tzOffset)
	if !ok {
		return "", fmt.Errorf("invalid schedule format: %s", schedule)
	}

	action := oasis.ScheduledAction{
		ID:              oasis.NewID(),
		Description:     p.Description,
		Schedule:        schedule,
		ToolCalls:       string(p.Tools),
		SynthesisPrompt: p.SynthesisPrompt,
		NextRun:         nextRun,
		Enabled:         true,
		CreatedAt:       now,
	}

	if err := t.store.CreateScheduledAction(ctx, action); err != nil {
		return "", err
	}

	return fmt.Sprintf("Scheduled: %s\nSchedule: %s\nNext run: %s",
		p.Description, schedule, formatLocalTime(nextRun, t.tzOffset)), nil
}

func (t *Tool) handleList(ctx context.Context) (string, error) {
	actions, err := t.store.ListScheduledActions(ctx)
	if err != nil {
		return "", err
	}
	if len(actions) == 0 {
		return "No scheduled actions.", nil
	}

	var out strings.Builder
	fmt.Fprintf(&out, "%d scheduled action(s):\n\n", len(actions))
	for i, a := range actions {
		status := "active"
		if !a.Enabled {
			status = "paused"
		}
		fmt.Fprintf(&out, "%d. %s [%s]\n   Schedule: %s | Next: %s\n",
			i+1, a.Description, status, a.Schedule,
			formatLocalTime(a.NextRun, t.tzOffset))
	}
	return out.String(), nil
}

func (t *Tool) handleUpdate(ctx context.Context, args json.RawMessage) (string, error) {
	var p struct {
		DescriptionQuery string  `json:"description_query"`
		Enabled          *bool   `json:"enabled"`
		Time             *string `json:"time"`
		Recurrence       *string `json:"recurrence"`
		Day              *string `json:"day"`
	}
	if err := json.Unmarshal(args, &p); err != nil {
		return "", fmt.Errorf("invalid args: %w", err)
	}

	matches, err := t.store.FindScheduledActionsByDescription(ctx, p.DescriptionQuery)
	if err != nil {
		return "", err
	}
	if len(matches) == 0 {
		return fmt.Sprintf("No scheduled action matching %q.", p.DescriptionQuery), nil
	}
	if len(matches) > 1 {
		names := make([]string, len(matches))
		for i, a := range matches {
			names[i] = a.Description
		}
		return fmt.Sprintf("Multiple matches: %s. Be more specific.", strings.Join(names, ", ")), nil
	}

	action := matches[0]
	var changes []string

	if p.Enabled != nil {
		if err := t.store.UpdateScheduledActionEnabled(ctx, action.ID, *p.Enabled); err != nil {
			return "", err
		}
		if *p.Enabled {
			changes = append(changes, "enabled")
		} else {
			changes = append(changes, "paused")
		}
	}

	if p.Time != nil || p.Recurrence != nil {
		// Parse current schedule
		parts := strings.SplitN(action.Schedule, " ", 2)
		currentTime := "08:00"
		currentRec := "daily"
		if len(parts) >= 1 {
			currentTime = parts[0]
		}
		if len(parts) >= 2 {
			currentRec = parts[1]
		}

		newTime := currentTime
		if p.Time != nil {
			newTime = *p.Time
		}

		newRec := currentRec
		if p.Recurrence != nil {
			day := ""
			if p.Day != nil {
				day = *p.Day
			}
			newRec = buildRecurrencePart(*p.Recurrence, day)
		}

		newSchedule := newTime + " " + newRec
		now := oasis.NowUnix()
		nextRun, ok := ComputeNextRun(newSchedule, now, t.tzOffset)
		if !ok {
			return "", fmt.Errorf("invalid schedule: %s", newSchedule)
		}

		action.Schedule = newSchedule
		action.NextRun = nextRun
		if err := t.store.UpdateScheduledAction(ctx, action); err != nil {
			return "", err
		}
		changes = append(changes, "schedule updated")
	}

	if len(changes) == 0 {
		return "No changes specified.", nil
	}

	return fmt.Sprintf("Updated %q: %s", action.Description, strings.Join(changes, ", ")), nil
}

func (t *Tool) handleDelete(ctx context.Context, args json.RawMessage) (string, error) {
	var p struct {
		DescriptionQuery string `json:"description_query"`
	}
	if err := json.Unmarshal(args, &p); err != nil {
		return "", fmt.Errorf("invalid args: %w", err)
	}

	if p.DescriptionQuery == "*" {
		count, err := t.store.DeleteAllScheduledActions(ctx)
		if err != nil {
			return "", err
		}
		return fmt.Sprintf("Deleted all %d scheduled action(s).", count), nil
	}

	matches, err := t.store.FindScheduledActionsByDescription(ctx, p.DescriptionQuery)
	if err != nil {
		return "", err
	}
	if len(matches) == 0 {
		return fmt.Sprintf("No scheduled action matching %q.", p.DescriptionQuery), nil
	}

	for _, a := range matches {
		if err := t.store.DeleteScheduledAction(ctx, a.ID); err != nil {
			return "", err
		}
	}

	if len(matches) == 1 {
		return fmt.Sprintf("Deleted: %s", matches[0].Description), nil
	}
	return fmt.Sprintf("Deleted %d scheduled action(s).", len(matches)), nil
}

// --- Schedule helpers ---

func buildScheduleString(timeStr, recurrence, day string) string {
	if timeStr == "" {
		timeStr = "08:00"
	}
	return timeStr + " " + buildRecurrencePart(recurrence, day)
}

func buildRecurrencePart(recurrence, day string) string {
	switch recurrence {
	case "once":
		return "once"
	case "custom":
		if day == "" {
			day = "monday,wednesday,friday"
		}
		return fmt.Sprintf("custom(%s)", normalizeDayList(day))
	case "weekly":
		if day == "" {
			day = "monday"
		}
		return fmt.Sprintf("weekly(%s)", strings.ToLower(strings.TrimSpace(day)))
	case "monthly":
		if day == "" {
			day = "1"
		}
		return fmt.Sprintf("monthly(%s)", day)
	default:
		return "daily"
	}
}

// ComputeNextRun calculates the next UTC timestamp for a schedule string.
// Schedule format: "HH:MM <recurrence>" where recurrence is:
// once, daily, custom(mon,wed,fri), weekly(monday), monthly(15)
func ComputeNextRun(schedule string, nowUnix int64, tzOffset int) (int64, bool) {
	parts := strings.SplitN(schedule, " ", 2)
	if len(parts) != 2 {
		return 0, false
	}

	timeParts := strings.Split(parts[0], ":")
	if len(timeParts) != 2 {
		return 0, false
	}
	hour := parseInt(timeParts[0])
	minute := parseInt(timeParts[1])
	if hour < 0 || hour > 23 || minute < 0 || minute > 59 {
		return 0, false
	}

	offsetSecs := int64(tzOffset) * 3600
	localNow := nowUnix + offsetSecs
	localDays := localNow / 86400
	localTimeOfDay := localNow % 86400
	targetTimeOfDay := int64(hour)*3600 + int64(minute)*60

	recurrence := strings.TrimSpace(parts[1])

	switch {
	case recurrence == "once" || recurrence == "daily":
		targetDay := localDays
		if localTimeOfDay >= targetTimeOfDay {
			targetDay++
		}
		localTS := targetDay*86400 + targetTimeOfDay
		return localTS - offsetSecs, true

	case strings.HasPrefix(recurrence, "custom("):
		daysStr := strings.TrimPrefix(recurrence, "custom(")
		daysStr = strings.TrimSuffix(daysStr, ")")
		currentDOW := ((localDays % 7) + 3) % 7 // Monday=0

		var bestAhead int64 = -1
		for _, dayName := range strings.Split(daysStr, ",") {
			targetDOW, ok := dayNameToDOW(strings.TrimSpace(dayName))
			if !ok {
				return 0, false
			}
			ahead := targetDOW - currentDOW
			if ahead < 0 {
				ahead += 7
			}
			if ahead == 0 && localTimeOfDay >= targetTimeOfDay {
				ahead = 7
			}
			if bestAhead < 0 || ahead < bestAhead {
				bestAhead = ahead
			}
		}
		if bestAhead < 0 {
			return 0, false
		}
		targetDay := localDays + bestAhead
		localTS := targetDay*86400 + targetTimeOfDay
		return localTS - offsetSecs, true

	case strings.HasPrefix(recurrence, "weekly("):
		dayName := strings.TrimPrefix(recurrence, "weekly(")
		dayName = strings.TrimSuffix(dayName, ")")
		targetDOW, ok := dayNameToDOW(dayName)
		if !ok {
			return 0, false
		}
		currentDOW := ((localDays % 7) + 3) % 7
		daysAhead := targetDOW - currentDOW
		if daysAhead < 0 {
			daysAhead += 7
		}
		if daysAhead == 0 && localTimeOfDay >= targetTimeOfDay {
			daysAhead = 7
		}
		targetDay := localDays + daysAhead
		localTS := targetDay*86400 + targetTimeOfDay
		return localTS - offsetSecs, true

	case strings.HasPrefix(recurrence, "monthly("):
		domStr := strings.TrimPrefix(recurrence, "monthly(")
		domStr = strings.TrimSuffix(domStr, ")")
		targetDOM := parseInt(domStr)
		if targetDOM < 1 || targetDOM > 31 {
			return 0, false
		}
		y, m, d := unixDaysToDate(localDays)
		targetY, targetM := y, m
		if int64(d) > int64(targetDOM) || (int64(d) == int64(targetDOM) && localTimeOfDay >= targetTimeOfDay) {
			if m == 12 {
				targetY = y + 1
				targetM = 1
			} else {
				targetM = m + 1
			}
		}
		targetDays := dateToUnixDays(targetY, targetM, targetDOM)
		localTS := targetDays*86400 + targetTimeOfDay
		return localTS - offsetSecs, true
	}

	return 0, false
}

func dayNameToDOW(name string) (int64, bool) {
	switch strings.ToLower(name) {
	case "monday", "mon", "senin":
		return 0, true
	case "tuesday", "tue", "selasa":
		return 1, true
	case "wednesday", "wed", "rabu":
		return 2, true
	case "thursday", "thu", "kamis":
		return 3, true
	case "friday", "fri", "jumat":
		return 4, true
	case "saturday", "sat", "sabtu":
		return 5, true
	case "sunday", "sun", "minggu":
		return 6, true
	}
	return 0, false
}

func normalizeDayList(input string) string {
	parts := strings.Split(input, ",")
	for i, p := range parts {
		parts[i] = strings.ToLower(strings.TrimSpace(p))
	}
	return strings.Join(parts, ",")
}

func parseInt(s string) int {
	n := 0
	for _, c := range s {
		if c >= '0' && c <= '9' {
			n = n*10 + int(c-'0')
		} else {
			return -1
		}
	}
	return n
}

func formatLocalTime(unix int64, tzOffset int) string {
	local := unix + int64(tzOffset)*3600
	days := local / 86400
	remainder := local % 86400
	hour := remainder / 3600
	minute := (remainder % 3600) / 60
	y, m, d := unixDaysToDate(days)
	return fmt.Sprintf("%04d-%02d-%02d %02d:%02d", y, m, d, hour, minute)
}

func unixDaysToDate(days int64) (year, month, day int) {
	// Algorithm from http://howardhinnant.github.io/date_algorithms.html
	z := days + 719468
	era := z / 146097
	if z < 0 {
		era = (z - 146096) / 146097
	}
	doe := z - era*146097
	yoe := (doe - doe/1460 + doe/36524 - doe/146096) / 365
	y := yoe + era*400
	doy := doe - (365*yoe + yoe/4 - yoe/100)
	mp := (5*doy + 2) / 153
	d := doy - (153*mp+2)/5 + 1
	m := mp + 3
	if mp >= 10 {
		m = mp - 9
	}
	if m <= 2 {
		y++
	}
	return int(y), int(m), int(d)
}

func dateToUnixDays(year, month, day int) int64 {
	y := int64(year)
	m := int64(month)
	d := int64(day)
	if m <= 2 {
		y--
	}
	era := y / 400
	if y < 0 {
		era = (y - 399) / 400
	}
	yoe := y - era*400
	var doy int64
	if m > 2 {
		doy = (153*(m-3)+2)/5 + d - 1
	} else {
		doy = (153*(m+9)+2)/5 + d - 1
	}
	doe := yoe*365 + yoe/4 - yoe/100 + doy
	return era*146097 + doe - 719468
}

File: crates/oasis-brain/src/tool/schedule.rs
use std::sync::Arc;

use async_trait::async_trait;
use oasis_core::error::{OasisError, Result};
use oasis_core::types::*;
use serde_json::json;

use crate::service::store::VectorStore;
use crate::util::{date_to_unix_days, format_due, unix_days_to_date};
use crate::tool::{Tool, ToolResult};

pub struct ScheduleTool {
    store: Arc<VectorStore>,
    tz_offset: i32,
}

impl ScheduleTool {
    pub fn new(store: Arc<VectorStore>, tz_offset: i32) -> Self {
        Self { store, tz_offset }
    }
}

#[async_trait]
impl Tool for ScheduleTool {
    fn definitions(&self) -> Vec<ToolDefinition> {
        vec![
            ToolDefinition {
                name: "schedule_create".to_string(),
                description: "Create a scheduled/recurring action that runs automatically. Use when the user wants something done periodically (daily briefings, recurring searches, regular summaries).".to_string(),
                parameters: json!({
                    "type": "object",
                    "properties": {
                        "description": { "type": "string", "description": "Human-readable description of what this scheduled action does" },
                        "time": { "type": "string", "description": "Time in HH:MM format (24-hour, user's local timezone)" },
                        "recurrence": { "type": "string", "enum": ["once", "daily", "custom", "weekly", "monthly"], "description": "How often to run. Use 'once' for one-time actions. Use 'custom' to pick specific days of the week." },
                        "day": { "type": "string", "description": "For weekly: single day name (monday/senin). For custom: comma-separated day names (e.g. 'senin,rabu,kamis' or 'monday,wednesday,friday'). For monthly: day number (1-31)." },
                        "tools": {
                            "type": "array",
                            "items": {
                                "type": "object",
                                "properties": {
                                    "tool": { "type": "string", "description": "Tool name: web_search, task_list, knowledge_search" },
                                    "params": { "type": "object", "description": "Parameters for the tool" }
                                },
                                "required": ["tool", "params"]
                            },
                            "description": "Tools to execute when the schedule fires"
                        },
                        "synthesis_prompt": { "type": "string", "description": "How to format/summarize results (e.g. 'Summarize in Indonesian, keep it brief')" }
                    },
                    "required": ["description", "time", "recurrence", "tools"]
                }),
            },
            ToolDefinition {
                name: "schedule_list".to_string(),
                description: "List all scheduled actions with their schedules, status, and next run time.".to_string(),
                parameters: json!({
                    "type": "object",
                    "properties": {}
                }),
            },
            ToolDefinition {
                name: "schedule_update".to_string(),
                description: "Update a scheduled action: enable/disable it or change its schedule.".to_string(),
                parameters: json!({
                    "type": "object",
                    "properties": {
                        "description_query": { "type": "string", "description": "Substring to match the scheduled action description" },
                        "enabled": { "type": "boolean", "description": "Set to true to enable, false to disable/pause" },
                        "time": { "type": "string", "description": "New time in HH:MM format (optional)" },
                        "recurrence": { "type": "string", "enum": ["once", "daily", "custom", "weekly", "monthly"], "description": "New recurrence (optional)" },
                        "day": { "type": "string", "description": "New day(s) — single for weekly, comma-separated for custom, number for monthly (optional)" }
                    },
                    "required": ["description_query"]
                }),
            },
            ToolDefinition {
                name: "schedule_delete".to_string(),
                description: "Delete a scheduled action. Matches by description substring, or '*' to delete all.".to_string(),
                parameters: json!({
                    "type": "object",
                    "properties": {
                        "description_query": { "type": "string", "description": "Substring to match the description, or '*' for all" }
                    },
                    "required": ["description_query"]
                }),
            },
        ]
    }

    async fn execute(&self, name: &str, args: &serde_json::Value) -> ToolResult {
        let result = match name {
            "schedule_create" => {
                let description = args["description"].as_str().unwrap_or("");
                let time = args["time"].as_str().unwrap_or("08:00");
                let recurrence = args["recurrence"].as_str().unwrap_or("daily");
                let day = args["day"].as_str();
                let tools = &args["tools"];
                let synthesis_prompt = args["synthesis_prompt"].as_str();
                self.handle_create(description, time, recurrence, day, tools, synthesis_prompt).await
            }
            "schedule_list" => self.handle_list().await,
            "schedule_update" => {
                let query = args["description_query"].as_str().unwrap_or("");
                let enabled = args["enabled"].as_bool();
                let time = args["time"].as_str();
                let recurrence = args["recurrence"].as_str();
                let day = args["day"].as_str();
                self.handle_update(query, enabled, time, recurrence, day).await
            }
            "schedule_delete" => {
                let query = args["description_query"].as_str().unwrap_or("");
                self.handle_delete(query).await
            }
            _ => return ToolResult::err(format!("Unknown schedule tool: {name}")),
        };
        match result {
            Ok(r) => ToolResult::ok(r),
            Err(e) => ToolResult::err(e.to_string()),
        }
    }
}

impl ScheduleTool {
    async fn handle_create(
        &self,
        description: &str,
        time: &str,
        recurrence: &str,
        day: Option<&str>,
        tools: &serde_json::Value,
        synthesis_prompt: Option<&str>,
    ) -> Result<String> {
        let schedule = match recurrence {
            "once" => format!("{time} once"),
            "custom" => {
                let d = day.unwrap_or("monday,wednesday,friday");
                let normalized = normalize_day_list(d);
                format!("{time} custom({normalized})")
            }
            "weekly" => {
                let d = day.unwrap_or("monday");
                format!("{time} weekly({d})")
            }
            "monthly" => {
                let d = day.unwrap_or("1");
                format!("{time} monthly({d})")
            }
            _ => format!("{time} daily"),
        };

        let now = now_unix();
        let next_run = compute_next_run(&schedule, now, self.tz_offset)
            .ok_or_else(|| OasisError::Config("invalid schedule format".to_string()))?;

        let action = ScheduledAction {
            id: new_id(),
            description: description.to_string(),
            schedule: schedule.clone(),
            tool_calls: normalize_tool_calls(tools),
            synthesis_prompt: synthesis_prompt.map(|s| s.to_string()),
            enabled: true,
            last_run: None,
            next_run,
            created_at: now,
        };

        self.store.insert_scheduled_action(&action).await?;

        let next_run_str = format_due(next_run, self.tz_offset);
        Ok(format!(
            "Scheduled: {description}\nSchedule: {schedule}\nNext run: {next_run_str}"
        ))
    }

    async fn handle_list(&self) -> Result<String> {
        let actions = self.store.list_scheduled_actions().await?;
        if actions.is_empty() {
            return Ok("No scheduled actions.".to_string());
        }

        let mut output = format!("{} scheduled action(s):\n\n", actions.len());
        for (i, a) in actions.iter().enumerate() {
            let status = if a.enabled { "active" } else { "paused" };
            let next = format_due(a.next_run, self.tz_offset);
            output.push_str(&format!(
                "{}. {} [{}]\n   Schedule: {} | Next: {}\n",
                i + 1,
                a.description,
                status,
                a.schedule,
                next,
            ));
        }
        Ok(output)
    }

    async fn handle_update(
        &self,
        query: &str,
        enabled: Option<bool>,
        time: Option<&str>,
        recurrence: Option<&str>,
        day: Option<&str>,
    ) -> Result<String> {
        let matches = self.store.find_scheduled_action_by_description(query).await?;
        if matches.is_empty() {
            return Ok(format!("No scheduled action matching \"{query}\"."));
        }
        if matches.len() > 1 {
            let names: Vec<_> = matches.iter().map(|a| a.description.as_str()).collect();
            return Ok(format!(
                "Multiple matches: {}. Be more specific.",
                names.join(", ")
            ));
        }

        let action = &matches[0];
        let mut changes = Vec::new();

        if let Some(en) = enabled {
            self.store.update_scheduled_action_enabled(&action.id, en).await?;
            changes.push(if en { "enabled" } else { "paused" });
        }

        if time.is_some() || recurrence.is_some() {
            let current_parts: Vec<&str> = action.schedule.splitn(2, ' ').collect();
            let new_time = time.unwrap_or(current_parts.first().copied().unwrap_or("08:00"));
            let new_rec = recurrence
                .map(|r| match r {
                    "once" => "once".to_string(),
                    "custom" => {
                        let d = day.unwrap_or("monday,wednesday,friday");
                        let normalized = normalize_day_list(d);
                        format!("custom({normalized})")
                    }
                    "weekly" => {
                        let d = day.unwrap_or("monday");
                        format!("weekly({d})")
                    }
                    "monthly" => {
                        let d = day.unwrap_or("1");
                        format!("monthly({d})")
                    }
                    _ => "daily".to_string(),
                })
                .unwrap_or_else(|| current_parts.get(1).copied().unwrap_or("daily").to_string());

            let new_schedule = format!("{new_time} {new_rec}");
            let now = now_unix();
            let next_run = compute_next_run(&new_schedule, now, self.tz_offset)
                .ok_or_else(|| OasisError::Config("invalid schedule".to_string()))?;
            self.store
                .update_scheduled_action_schedule(&action.id, &new_schedule, next_run)
                .await?;
            changes.push("schedule updated");
        }

        if changes.is_empty() {
            return Ok("No changes specified.".to_string());
        }

        Ok(format!(
            "Updated \"{}\": {}",
            action.description,
            changes.join(", ")
        ))
    }

    async fn handle_delete(&self, query: &str) -> Result<String> {
        if query == "*" {
            let count = self.store.delete_all_scheduled_actions().await?;
            return Ok(format!("Deleted all {count} scheduled action(s)."));
        }

        let matches = self.store.find_scheduled_action_by_description(query).await?;
        if matches.is_empty() {
            return Ok(format!("No scheduled action matching \"{query}\"."));
        }

        let mut deleted = 0;
        for a in &matches {
            self.store.delete_scheduled_action(&a.id).await?;
            deleted += 1;
        }

        if deleted == 1 {
            Ok(format!("Deleted: {}", matches[0].description))
        } else {
            Ok(format!("Deleted {deleted} scheduled action(s)."))
        }
    }
}

// ─── Schedule utility functions ───────────────────────────────────────

/// Normalize tool_calls JSON: if the LLM produced string-encoded objects,
/// parse and re-serialize as proper objects.
fn normalize_tool_calls(tools: &serde_json::Value) -> String {
    if let Some(arr) = tools.as_array() {
        if arr.iter().all(|v| v.is_string()) {
            let parsed: Option<Vec<serde_json::Value>> = arr
                .iter()
                .map(|v| serde_json::from_str(v.as_str().unwrap_or("")))
                .collect::<std::result::Result<Vec<_>, _>>()
                .ok();
            if let Some(objects) = parsed {
                return serde_json::Value::Array(objects).to_string();
            }
        }
    }
    tools.to_string()
}

/// Compute the next run time (UTC) for a schedule string.
pub fn compute_next_run(schedule: &str, now: i64, tz_offset: i32) -> Option<i64> {
    let parts: Vec<&str> = schedule.splitn(2, ' ').collect();
    if parts.len() != 2 {
        return None;
    }

    let time_parts: Vec<&str> = parts[0].split(':').collect();
    if time_parts.len() != 2 {
        return None;
    }
    let hour: i64 = time_parts[0].parse().ok()?;
    let minute: i64 = time_parts[1].parse().ok()?;
    if hour > 23 || minute > 59 {
        return None;
    }

    let offset_secs = (tz_offset as i64) * 3600;
    let local_now = now + offset_secs;
    let local_days = local_now / 86400;
    let local_time_of_day = local_now % 86400;
    let target_time_of_day = hour * 3600 + minute * 60;

    let recurrence = parts[1].trim();

    match recurrence {
        "once" | "daily" => {
            let target_day = if local_time_of_day >= target_time_of_day {
                local_days + 1
            } else {
                local_days
            };
            let local_ts = target_day * 86400 + target_time_of_day;
            Some(local_ts - offset_secs)
        }
        s if s.starts_with("custom(") => {
            let days_str = s.trim_start_matches("custom(").trim_end_matches(')');
            let current_dow = ((local_days % 7) + 3) % 7;
            let mut best_ahead: Option<i64> = None;

            for day_name in days_str.split(',') {
                let target_dow = day_name_to_dow(day_name.trim())?;
                let mut ahead = target_dow - current_dow;
                if ahead < 0 {
                    ahead += 7;
                }
                if ahead == 0 && local_time_of_day >= target_time_of_day {
                    ahead = 7;
                }
                best_ahead = Some(match best_ahead {
                    Some(b) if b <= ahead => b,
                    _ => ahead,
                });
            }

            let days_ahead = best_ahead?;
            let target_day = local_days + days_ahead;
            let local_ts = target_day * 86400 + target_time_of_day;
            Some(local_ts - offset_secs)
        }
        s if s.starts_with("weekly(") => {
            let day_name = s.trim_start_matches("weekly(").trim_end_matches(')');
            let target_dow = day_name_to_dow(day_name)?;
            let current_dow = ((local_days % 7) + 3) % 7;
            let mut days_ahead = target_dow - current_dow;
            if days_ahead < 0 {
                days_ahead += 7;
            }
            if days_ahead == 0 && local_time_of_day >= target_time_of_day {
                days_ahead = 7;
            }
            let target_day = local_days + days_ahead;
            let local_ts = target_day * 86400 + target_time_of_day;
            Some(local_ts - offset_secs)
        }
        s if s.starts_with("monthly(") => {
            let dom_str = s.trim_start_matches("monthly(").trim_end_matches(')');
            let target_dom: i64 = dom_str.parse().ok()?;
            if target_dom < 1 || target_dom > 31 {
                return None;
            }
            let (y, m, d) = unix_days_to_date(local_days);
            let (target_y, target_m) =
                if d > target_dom || (d == target_dom && local_time_of_day >= target_time_of_day) {
                    if m == 12 {
                        (y + 1, 1)
                    } else {
                        (y, m + 1)
                    }
                } else {
                    (y, m)
                };
            let target_days = date_to_unix_days(target_y, target_m, target_dom);
            let local_ts = target_days * 86400 + target_time_of_day;
            Some(local_ts - offset_secs)
        }
        _ => None,
    }
}

/// Normalize comma-separated day names to lowercase, trimmed.
fn normalize_day_list(input: &str) -> String {
    input
        .split(',')
        .map(|d| d.trim().to_lowercase())
        .collect::<Vec<_>>()
        .join(",")
}

fn day_name_to_dow(name: &str) -> Option<i64> {
    match name.to_lowercase().as_str() {
        "monday" | "mon" | "senin" => Some(0),
        "tuesday" | "tue" | "selasa" => Some(1),
        "wednesday" | "wed" | "rabu" => Some(2),
        "thursday" | "thu" | "kamis" => Some(3),
        "friday" | "fri" | "jumat" => Some(4),
        "saturday" | "sat" | "sabtu" => Some(5),
        "sunday" | "sun" | "minggu" => Some(6),
        _ => None,
    }
}
